### PR TITLE
Fully process always_inline (second attempt)

### PR DESCRIPTION
**Pull request by tautschnig**
This was previously proposed and merged in #1510, but had to be reverted in #1527. #1519 then provided an additional test, which is now included (together with the fix).

### CODE REVIEW

File: regression/ansi-c/always_inline1/main.c [hunk 1]
```diff
@@ -0,0 +1,37 @@
+// this is a GCC extension
+
+#ifdef __GNUC__
+static inline __attribute__((always_inline)) _Bool
```
Comment (smowton) [inline regression/ansi-c/always_inline1/main.c:4]: Could you minimise this test-case so it's clearer which parts are crucial to the test and which is unrelated stuff the kernel's RCU code happens to do? If you want to keep the real-world example you could include this as a block comment, then use a simplified version for the actual test?

File: regression/ansi-c/always_inline1/test.desc [hunk 2]
```diff
@@ -0,0 +1,8 @@
+CORE
+main.c
+
+^EXIT=0$
+^SIGNAL=0$
+--
+^warning: ignoring
+^CONVERSION ERROR$
```
Comment (smowton) [inline regression/ansi-c/always_inline1/test.desc:8]: Add a comment to the test specifying *why* we expect an error

File: src/ansi-c/c_typecheck_expr.cpp [hunk 3]
```diff
@@ -1945,6 +1949,73 @@ void c_typecheck_baset::typecheck_side_effect_function_call(
       warning().source_location=f_op.find_source_location();
       warning() << "function `" << identifier << "' is not declared" << eom;
     }
+    else if(
+      sym_entry->second.type.get_bool(ID_C_inlined) &&
+      sym_entry->second.is_macro && sym_entry->second.value.is_not_nil())
+    {
+      // calling a function marked as always_inline
```
Comment (smowton) [inline src/ansi-c/c_typecheck_expr.cpp:1972]: Or, presumably, a macro? Or is the `is_macro` flag only used for functions that are not in fact macros?

File: src/ansi-c/c_typecheck_expr.cpp [hunk 4]
```diff
@@ -1945,6 +1949,73 @@ void c_typecheck_baset::typecheck_side_effect_function_call(
...
+      replace_symbolt replace;
+
+      const code_typet::parameterst &parameters = func_type.parameters();
+      auto p_it = parameters.begin();
+      for(const auto &arg : expr.arguments())
+      {
+        if(p_it == parameters.end())
+        {
+          err_location(f_op);
+          error() << "function call has additional arguments, "
```
Comment (smowton) [inline src/ansi-c/c_typecheck_expr.cpp]: More arguments than parameters, i.e. varargs?

File: src/ansi-c/c_typecheck_expr.cpp [hunk 5]
```diff
@@ -1945,6 +1949,73 @@ void c_typecheck_baset::typecheck_side_effect_function_call(
...
+        ID_statement_expression, func_type.return_type());
+      body.make_block();
+
+      // simulates parts of typecheck_function_body
+      typet cur_return_type = return_type;
+      return_type = func_type.return_type();
+      typecheck_code(body);
+      return_type.swap(cur_return_type);
+
+      // replace final return by an ID_expression
```
Comment (smowton) [inline src/ansi-c/c_typecheck_expr.cpp:2023]: Could there be other returns? If so what happens?

File: src/ansi-c/c_typecheck_expr.cpp [hunk 6]
```diff
@@ -1945,6 +1949,73 @@ void c_typecheck_baset::typecheck_side_effect_function_call(
...
+      if(p_it != parameters.end())
+      {
+        err_location(f_op);
+        error() << "function call has missing arguments, "
+                << "cannot apply always_inline" << eom;
+        throw 0;
+      }
+
+      codet body = to_code(func_sym.value);
+      replace(body);
```
Comment (smowton) [inline src/ansi-c/c_typecheck_expr.cpp:2011]: So this will replace formal parameters with actuals, but not actually cause locals to alias because they retain the qualified names (`f::1::x` vs. `g::1::x`) they had originally? That seems right for inlining but wrong for macros?

**Review by smowton — COMMENTED** [review]
Some questions inline. Also, why do we have this at all? Surely always-inline is an optimisation that we could reasonably ignore?

**Comment by smowton** [comment]
Ah, so this is because __builtin_constant_p expands to some assertion which we were spuriously failing? Could we replace the builtin with something symex interprets instead?

**Comment by smowton** [comment]
Gah, sorry, I understand now -- this code is trying to cause a compile error if it can't prove something is constant?

**Comment by tautschnig** [comment]
> Gah, sorry, I understand now -- this code is trying to cause a compile error if it can't prove something is constant?

Yes, I believe that's the way that some code in the Linux kernel uses it.

**Comment by tautschnig** [comment]
@smowton Thanks a lot for all the comments, I will carefully review my code to come up with proper answers or fixes.

File: regression/ansi-c/always_inline1/main.c (hunk 1 above)
Comment (tautschnig) [reply regression/ansi-c/always_inline1/main.c:4]: Actually everything is necessary here, but I'll add comments to explain why/what it is testing.

File: src/ansi-c/c_typecheck_expr.cpp (hunk 3 above)
Comment (tautschnig) [reply src/ansi-c/c_typecheck_expr.cpp:1972]: Yes, `is_macro` isn't ever used with (preprocessor) macros.

File: src/ansi-c/c_typecheck_expr.cpp (hunk 5 above)
Comment (tautschnig) [reply src/ansi-c/c_typecheck_expr.cpp:2023]: It's unsupported - now it's being tested for and reported.

File: src/ansi-c/c_typecheck_expr.cpp (hunk 6 above)
Comment (tautschnig) [reply src/ansi-c/c_typecheck_expr.cpp:2011]: As said above, we don't actually do preprocessor macros in this way, so it should be fine.

File: regression/ansi-c/always_inline1/test.desc (hunk 2 above)
Comment (tautschnig) [reply regression/ansi-c/always_inline1/test.desc:8]: Done.

File: src/ansi-c/ansi_c_declaration.cpp [hunk 7]
```diff
@@ -164,6 +166,9 @@ void ansi_c_declarationt::to_symbol(
           symbol.is_extern=false;
         else  if(get_is_extern()) // traditional GCC
           symbol.is_file_local=true;
+
+        if(get_is_always_inline())
+          symbol.is_macro = true;
```
Comment (smowton) [inline src/ansi-c/ansi_c_declaration.cpp:171]: Can we please rename this then? It's really confusing that things annotated `is_macro` and things that are macros are disjoint (!)

**Review by smowton — APPROVED** [review]
This looks good except that using "macro" to refer to things that are not macros, and never to refer to things that are macros, is maddening. Please change this if at all possible.

File: src/ansi-c/ansi_c_declaration.cpp (hunk 7 above)
Comment (smowton) [reply src/ansi-c/ansi_c_declaration.cpp:171]: I also note we have `symbolt::is_function` which returns false if `is_macro` is set!

Comment (tautschnig) [reply src/ansi-c/ansi_c_declaration.cpp:171]: Upon further thought I'm inclined to say that `is_macro` actually is the right terminology for all the current uses: all of them (typedefs, enum values, linker aliases, always-inline functions) are names with some value (which might be code) behind them and need to be expanded in place in order to evaluate the expression or statement they are involved in. We currently consume code after the C/C++ preprocessor has been run, so those kinds of "macros" are not currently seen - but actually, if they were, they would fit the very same definition and would likely be marked as such.

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 688a0ab2308c0dc1f1030b64d08f3ea38fa5eeb3).

**Review by kroening — APPROVED** [review]
